### ISSUE TITLE
Bug?: dictionary_decompounder filter ignores marked keywords

### ISSUE THREAD

**Issue by iamjochem** [elastic/elasticsearch#21034]
The following looks like a bug to me (but maybe I misunderstand the intent of the two filters in question and how they are meant to interact).

**Elasticsearch version** | 2.2
---------------------------- | ----
**Plugins installed** |  https://github.com/francesconero/elasticsearch-concatenate-token-filter/tree/2.2.0
**JVM version** | java version "1.8.0_51", Java(TM) SE Runtime Environment (build 1.8.0_51-b16), Java HotSpot(TM) 64-Bit Server VM (build 25.51-b03, mixed mode)
**OS version** | Mac OSX 10.10.5

## The problem:

 I am trying to use the `dictionary_decompounder` filter to extract sub-words and at the same time stop other words from being stemmed/decompounded using the `keyword_marker` filter - the `dictionary_decompounder` ignores tokens marked as keywords and de-compounds them. 

The expectation is that the tokens marked as keywords are not touched by the `dictionary_decompounder` filter (in the same way as stemmer filters do not touch them). 

Additionally I notice that the `dictionary_decompounder` will generate duplicate tokens when the input token is equal to a matching sub-word ("keuken" in the example below) - this looks like it is not intended but it is easily worked around using a `unique` filter (so, for me, it's not so much of a problem)

### 1. create a test index:

>_N.B._ the `control_*` analyzers are included to show that stemming does not occur on keywords and also that the word "messing" is [incorrectly] stemmed to "mes" 

```json
{
    "analysis"    : {
        "analyzer"      : {
            "control_stem"      : {
                "type"          : "custom",
                "tokenizer"     : "standard",
                "filter"        : ["kp_snow"]
            },

            "control_kwstem"    : {
                "type"          : "custom",
                "tokenizer"     : "standard",
                "filter"        : ["nl_keywords", "kp_snow"]
            },

            "test"              : {
                "type"          : "custom",
                "tokenizer"     : "standard",
                "filter"        : ["nl_keywords", "nl_dehyphenate", "kp_snow"]
            }
        },

        "filter"        : {
            "kp_snow"           : {
                "type"                      : "snowball",
                "language"                  : "Kp"
            },


            "nl_keywords"       : {
                "type"                      : "keyword_marker",
                "keywords"                  : [
                    "messing"
                ]
            },

            "nl_dehyphenate": {
                "type"                      : "dictionary_decompounder",
                "min_word_size"             : 6,
                "min_subword_size"          : 3,
                "max_subword_size"          : 15,
                "word_list"                 : [
                    "keuken",
                    "mes"
                ]
            }
        }
    }
}
```

### 2. Perform an analyzer query 
Assuming the created index is called "**decomp_test**" the following request illustrate expected/actual analyzer results:

```sh
http://localhost:9200/decomp_test/_analyze?analyzer=test
```

Actual Tokens | Expected Tokens
---------------- | --------------------
 messing, mes, keukenme, keuk, mes, keuk, keuk | messing, keukenme, keuk, mes, keuk

**Comment by jimferenczi** [comment]
This works as expected, the `keyword_marker` protects words from being modified by stemmers only: 
https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-keyword-marker-tokenfilter.html
The `dictionary_decompounder` is not a stemmer so it ignores the keyword marker.
I'd consider opening an issue in Lucene (the decompounder is in Lucene) to see if it would be possible to change the behavior of this filter.

**Comment by iamjochem** [comment]
@jimferenczi - thanks for the clarification, bit disappointing but understandable!

As a non-expert in linguistics I read "Keyword Marker Token Filter" as "this filter marks tokens so other filters don't do their magic on them" ... the part about it only applying to stemmers gets forgotten (even though I did actually read that some time ago!).

I don't think I'll bother with trying to open an issue with the Lucene project - I know nothing about the details of Lucene and it's nomenclature and if I try to explain my problems in terms what I am seeing in ElasticSearch (which is all I am able to do) I will be [rightly] ignored.

I'll will have to give up on using the decompounder (the `hyphenation_decompounder` does not work at all either ([see this closed issue, where I also commented regarding the fact that the problem does not look solved](https://github.com/elastic/elasticsearch/issues/13935))

Thank you again for taking the time to read & reply!
